### ISSUE TITLE
Animation events broken for Clamped WrapMode

### ISSUE THREAD

**Issue by ataulien** [GameFoundry/bsf#303]
There seems to be a problem with how animation events are handled when an animation is *clamped* instead of *wrapped*, as they only seem to be triggered during the first few seconds the application is running. They are also being triggered even though the animation is at the wrong time.

I've put a debug-log print at the animaiton event callback like this:
```cpp
    bs::AnimationClipState state;
    mSubAnimation->getState(clip, state);

    gDebug().logDebug(bs::StringUtil::format(
        "[VisualCharacter] Got animation event: {0}:{1} while playing {2} at {3}", command, action,
        clip->getName(), state.time));

    gDebug().logDebug("Animation has the following events: ");
    for (auto& event : clip->getEvents())
    {
      gDebug().logDebug(bs::StringUtil::format(" - {0}: {1}", event.time, event.name));
    }
```

Which gives me the following output:
```
[DEBUG] [VisualCharacter] Got animation event: PLAYCLIP:S_RUN while playing HUMANS-S_RUN at 0.110297
[DEBUG] Animation has the following events: 
[DEBUG]  - 1.955000: PLAYCLIP:S_RUN
```

As you can see, the even is triggered at a time of `0.110297`, even though the only event is supposed to be triggered at `1.955000`.

## Problem 1 - LastFrameTime is not Last Frame Time

I dug a bit deeper and might have found two problems with the current implementation. First of all, the parameters of `Animation::triggerEvents(float lastFrameTime, float delta)` seem fishy.

I put a breakpoint inside the function and started the debugger. As expected, I got `lastFrameTime` and `delta` as 0, as the breakpoint hit in the first frame. Then, once the breakpoint hit in the second frame I got the exact time I took to press the *run* button for *both* of them!

I expected `lastFrameTime` to be 0 and `delta` to be the time since the first frame.

Turns out, `lastFrameTime` is always fed the *current frame time*! This can be seen in `AnimationManager::update`:

```
float timeDelta = mAnimationTime - mLastAnimationUpdateTime;
mLastAnimationUpdateTime = mAnimationTime;

// Trigger events and update attachments (for the data from the last frame)
if(async)
{
	for (auto& anim : mAnimations)
	{
		anim.second->updateFromProxy();
		anim.second->triggerEvents(mAnimationTime, gTime().getFrameDelta());
	}
}
```

should rather be this:

```cpp
float timeDelta = mAnimationTime - mLastAnimationUpdateTime;

// Trigger events and update attachments (for the data from the last frame)
if(async)
{
	for (auto& anim : mAnimations)
	{
		anim.second->updateFromProxy();
		anim.second->triggerEvents(mLastAnimationUpdateTime, gTime().getFrameDelta());
	}
}

mLastAnimationUpdateTime = mAnimationTime;
```
Note that I changed the parameter of `triggerEvents` and moved the assign-operation further down.

## Problem 2 - WrapTime gets the global running time

The second problem is inside `Animation::triggerEvents` and I'm not quite sure on how to best fix it. Since I learned that the parameter `lastFrameTime` is a global running time, clamping it here doesn't make sense:
```cpp
bool loop = clipInfo.state.wrapMode == AnimWrapMode::Loop;

float start = lastFrameTime;
float end = start + delta;
float clipLength = clipInfo.clip->getLength();

float wrappedStart = start;
float wrappedEnd = end;
AnimationUtility::wrapTime(wrappedStart, 0.0f, clipLength, loop);
AnimationUtility::wrapTime(wrappedEnd, 0.0f, clipLength, loop);
```

Once `lastFrameTime` is higher than the clip length, `wrappedStart` and `wrappedEnd` will be always clamped to the clip length, no matter when the animation was actually started!

There is no way to trigger any more events on clamped animations after the first few seconds of runtime. A possible solution would be to subtract the time when the clip was started from `lastFrameTime` first, but I'm not sure on how to get that or whether there would be a better solution.

**Comment by BearishSun** [comment]
Thanks for investigating so much detail. I'm currently working on a bigger task which might not be done for a few days, so I'll take a closer look then.

**Comment by BearishSun** [comment]
I had the time to take a short look at this today, and indeed there were multiple problems with how time was handled. I whipped up a potential solution but didn't have time to test it yet. If you get a chance try applying the patch below and see if that fixes the problems.

Patch: https://www.dropbox.com/s/d0h4jsud1rpz64n/animation.patch?dl=0

(Put the patch file in your `bsf` root directory and run `git apply animation.patch`.)

**Comment by ataulien** [comment]
Thanks for the patch, I'll try it later when I find some time!

**Comment by BearishSun** [comment]
I submitted a fix in a7114140b4cc98d459eb53d83ee70130f028dcc9. In my test it seems to work but if you're still having problems feel free to reopen the issue.

**Comment by ataulien** [comment]
I can confirm that it is indeed working fine now! Thank you!
